pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1383,13 +1383,21 @@ impl RemoteTimelineClient {
 
             let upload_result: anyhow::Result<()> = match &task.op {
                 UploadOp::UploadLayer(ref layer, ref layer_metadata) => {
-                    let path = layer.local_path();
+                    let local_path = layer.local_path();
+
+                    let remote_path = remote_layer_path(
+                        &self.tenant_shard_id.tenant_id,
+                        &self.timeline_id,
+                        layer_metadata.shard,
+                        &layer.layer_desc().filename(),
+                        layer_metadata.generation,
```

Comment:
This is using `layer_metadata.generation` for the remote path whereas previously we were using `self.generation`.

I'm pretty sure that's wrong.

Please re-check all other such converstions, I spotted this kind of by accident.
(Also wondering if there's a way to prevent such things through better code structure / type system.)

Reply:
This should be equivalent (we only upload layers created in the current generation), but I'm adding an assertion and re-testing to make sure.

Reply:
Assertion in 825a52016d5c551bcba5245c786a0d53211be6c6